Create a way for admins to copy courses from a production instance of the Dashboard

enelesmai opened the pull request:
NOTE: Please review the pull request process before opening your first PR: https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/master/CONTRIBUTING.md#pull-request-process

Issue: #4911 

## What this PR does
We have a script for copying the core data of one program on a public instance of the Dashboard to another instance: `setup/copy_course_from_production.rb`. However this can only be used manually from a rails console.

A good step forward would be to add a controller and very simple frontend for this functionality, so that it's really easy for a developer to copy a course from production for testing.

This will also require a thorough refactor of the script into something that looks more like one of the `app/services `files, along with comprehensive tests for the controller and service.

## Screenshots
Before:
<img width="1052" alt="image" src="https://user-images.githubusercontent.com/5160907/163512102-1ba87aec-bb47-4df0-9f1c-8ec3e30629da.png">

After:
<img width="912" alt="image" src="https://user-images.githubusercontent.com/5160907/163511929-3bc68add-b5de-486d-b563-e2a363522736.png">

<img width="1150" alt="image" src="https://user-images.githubusercontent.com/5160907/163511970-4e6ae9fb-48a8-4324-bc5e-0db985ec111c.png">

** Usage:

Copy the url to the text box:
![image](https://user-images.githubusercontent.com/5160907/164531694-85397e31-8613-4c76-84d6-4136f15cc13e.png)

Result:
![image](https://user-images.githubusercontent.com/5160907/164532106-f536ed34-d912-45bb-8ffc-0befa89a4dc0.png)

![image](https://user-images.githubusercontent.com/5160907/164532316-1ee64b20-5229-487b-b8a6-86e2658f9619.png)


## Open questions and concerns

* Run added test cases
**`rspec ./spec/services/copy_course_from_production_spec.rb`
**`rspec ./spec/controllers/copy_course_from_production_controller_spec.rb`

An url example is added to test. VCR.use_cassette is used to perform HTTP request. The test cases expect to validate the information is loaded from production and copied to the local db for test.

>>

enelesmai commented:
I have a question here @ragesoss, for the frontend is it ok only adding a textbox input for the url? and then a button to execute the service? or do you suggest something else?

ragesoss commented:
Yes, a simple text input is fine.

On Fri, Apr 8, 2022, 6:47 PM Xóchitl Selene Flores Pérez <
***@***.***> wrote:

> I have a question here @ragesoss <https://github.com/ragesoss>, for the
> frontend is it ok only adding a textbox input for the url? and then a
> button to execute the service? or do you suggest something else?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/WikiEducationFoundation/WikiEduDashboard/pull/4914#issuecomment-1093573709>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGPEY5HNOIFHERNVZOBF63VEDOSRANCNFSM5SYF3NTQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>

enelesmai commented:
Hello @ragesoss  I have one more question: Where should I put the link to access to this page?

ragesoss commented:
I guess it should link from /admin .

enelesmai commented:
> I guess it should link from /admin .

I was exploring and found that each course have a list of actions, and there is a similar functionality called ´Clone course´. I think I can add the option to copy the course on this list, and do the necessary changes so we don't have to use the url to copy the course and so this functionality will be easier to use. What do you think about it?

![image](https://user-images.githubusercontent.com/5160907/163026161-69c4862a-1d75-4615-b57f-92624f6f4cfc.png)

ragesoss commented:
What you're showing in this screenshot is the 'Actions' buttons for an
individual course. To use this, you would need to first set up a course
that matched one on production. It will be much easier to use a URL to copy
a course, without already having a course page for it in your dev
environment.

-Sage

On Tue, Apr 12, 2022 at 11:08 AM Xóchitl Selene Flores Pérez <
***@***.***> wrote:

> I guess it should link from /admin .
>
> I was exploring and found that each course have a list of actions, and
> there is a similar functionality called ´Clone course´. I think I can add
> the option to copy the course on this list, and do the necessary changes so
> we don't have to use the url to copy the course and so this functionality
> will be easier to use. What do you think about it?
>
> [image: image]
> <https://user-images.githubusercontent.com/5160907/163026161-69c4862a-1d75-4615-b57f-92624f6f4cfc.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/WikiEducationFoundation/WikiEduDashboard/pull/4914#issuecomment-1097035138>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGPEY472QK2EDPJMVXYKKTVEW327ANCNFSM5SYF3NTQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>

enelesmai commented:
> 

That's right I was thinking about that, it was just an idea but it will need a lot of changes like the way we get the course data and the validation to show this option only to admins and to production courses, in addition to it could be used in production only. So I'm just adding the new page, hehe. ✌

enelesmai commented:
Hello @ragesoss, I opened this PR for review

ragesoss commented on app/controllers/copy_course_from_production_controller.rb:
```diff
@@ -0,0 +1,19 @@
+
+  def copy
+    copied_course = CopyCourseFromProduction.new(params[:url])
+    pp "http://localhost:3000/courses/#{copied_course.course_slug}"
```
We shouldn't have any `pp` output in the production rails code.

enelesmai replied:
Removed here [ecd6486](https://github.com/WikiEducationFoundation/WikiEduDashboard/pull/4914/commits/ecd6486288eafd4dff4518664b57ded7d7360153)

ragesoss commented on app/views/admin/index.html.haml:
```diff
@@ -20,6 +20,8 @@
     %li
       %a{href: 'https://dashboard-testing.wikiedu.org/rails/mailers'} Email templates
+    %li
+      %a{href: '/course_copy_prod'} Copy course from production
```
Probably a better label for this will be "Copy course from another Dashboard server"

enelesmai replied:
Updated here [82d04a6](https://github.com/WikiEducationFoundation/WikiEduDashboard/pull/4914/commits/82d04a6c1dbb8c370b5f958e0134a994b43b98ef)

ragesoss commented on spec/controllers/copy_course_from_production_controller_spec.rb:
```diff
@@ -0,0 +1,40 @@
+            it 'copies the course' do
+                subject
+                expect(response.status).to eq(200)
+                expect(Course.exists?(:title => existent_prod_course)).to eq(true)
```
This doesn't make sense, as `existent_prod_course` is a URL string, not a course title (which in this case, is "Invertebrate Zoology").

ragesoss replied:
This probably also needs to stub `CopyCourseFromProduction`, as we don't want to hit a production server more than once at most in the test suite. (We don't currently hit production dashboard servers at all in the test suite, and doing so makes me just a little bit uneasy, but I think it makes sense for a test of the service itself.)

enelesmai replied:
I did the change using slug property to check the existence [d75a150](https://github.com/WikiEducationFoundation/WikiEduDashboard/pull/4914/commits/d75a1504d60105e3a1ad5da485c735e90c16cd37)

enelesmai replied:
I moved the test case to check service only from CopyCourseFromProduction [d49f462](https://github.com/WikiEducationFoundation/WikiEduDashboard/pull/4914/commits/d49f46200acd3cf91310105c9ff01dd8c7c77940)

ragesoss commented on spec/services/copy_course_from_production_spec.rb:
```diff
@@ -0,0 +1,19 @@
+      described_class.new(existent_prod_course)
+    end
+
+    it 'copy course to dev env' do
```
This should test more than just the returned `result`. It should verify that examples of each of the important types of data got copied over as well: students, instructors, and assignments.

enelesmai replied:
Do you think this kind of examples are enough? [6e16b75](https://github.com/WikiEducationFoundation/WikiEduDashboard/pull/4914/commits/6e16b75e308a948ef00753b658878b351947d7b6)

ragesoss replied:
Looks good.

ragesoss commented on spec/services/copy_course_from_production_spec.rb:
```diff
@@ -0,0 +1,34 @@
+
+  context 'with courses in production' do
+    let(:subject) do
+      described_class.new(url_base + existent_prod_course_slug)
```
Since these tests make external web requests from the server, you need to use VCR to allow those requests to go through in the test environment.

enelesmai replied:
@ragesoss I just realize about this comment, in the morning I added de VCR.use_cassette block to each test case, do you think is ok there?

ragesoss replied:
The way you did it should be fine.

ragesoss commented:
Are these specs passing for you locally?

enelesmai commented:
> Are these specs passing for you locally?
I had my environment kind of broken but right now I already fixed it, I added some fixes but the specs are not showing anything 🤔

<img width="1008" alt="image" src="https://user-images.githubusercontent.com/5160907/164535650-37e09c69-de2a-4a15-ab45-bcbca5df2acc.png">

ragesoss commented:
>  I had my environment kind of broken but right now I already fixed it, I added some fixes but the specs are not showing anything thinking
> 
> <img alt="image" width="1008" src="https://user-images.githubusercontent.com/5160907/164535650-37e09c69-de2a-4a15-ab45-bcbca5df2acc.png">

Hmm... weird. I haven't tried running these locally yet, but they have been failing in CI. But I would expect them to return something when you run them locally, just like they do on CI.

enelesmai commented:
Only one try more, I'm still fixing my rspec installation 😰

ragesoss commented:
@enelesmai the one failing test is unrelated to this PR, I think. Is this PR ready to go from your perspective, or are there other changes you're still working on?
